src/main/scala/analysis/IrreducibleLoops.scala

```diff
@@ -110,167 +197,145 @@ class LoopDetector(cfg: Program) {
-        if (nodeDFSPpos(b) > 0) {
-          // Case (b)
-          // b is in DFSP(b0)
-          headers += b
-
-          val newLoop = if (loops.contains(b)) loops(b) else Loop(b)
-          edgeStack.reverse.slice(nodeDFSPpos(b) - 1, nodeDFSPpos(b0)).foreach { pEdge => newLoop.addEdge(pEdge) }
-          newLoop.backEdges += edge
-
-          // Add loop entry edge
-          IntraProcIRCursor.pred(b).foreach { predNode =>
-            val predEdge = LoopEdge(predNode, b)
-            if (nodeDFSPpos.contains(predNode)) {
-              if (nodeDFSPpos(predNode) > 0 && predEdge != edge) {
-                newLoop.entryEdges += predEdge
-              }
-            }
-          }
-
-          if (!loops.contains(b)) {
-            loops(b) = newLoop
-          }
-
-          tag_lhead(b0, b)
-        } else if (!iloopHeaders.contains(b)) {
-          // Case (c) - do nothing
-          // b is not part of this path, and it's not a header, so we can ignore it.
-          // (in effect we add it to our path later, but we do that when we discover a new header
-          //  by looking through the instruction call stack)
-        } else {
-          var h: CFGPosition = iloopHeaders(b)
-          if (nodeDFSPpos(h) > 0) {
-            // Case (d)
-            // h is in DFSP(b0)
-            val loop = loops(h)
-
-            // Add current path to the existing loop (a new path in the loop is discovered)
-            // This can happen for example in the case that there is a branch in a loop, or a `continue` stmt, etc
-            edgeStack.reverse.slice(nodeDFSPpos(h) - 1, nodeDFSPpos(b0)).foreach {
-              pEdge => loop.addEdge(pEdge)
-            }
-            IntraProcIRCursor.succ(b).filter(n => n == h).foreach { n =>
-              val outEdge = LoopEdge(b, n)
-              loop.addEdge(outEdge)
-            }
-
-            tag_lhead(b0, h)
-          } else {
-            // Case (e)
-            // reentry (irreducible!)
-            val loop = loops(h)
-            loop.reducible = false
-            loop.reentries += edge
-
-            // Make outer loops irreducible if the originating node of the re-entry edge is not within those loops
-            IntraProcIRCursor.succ(b).foreach { nextNode =>
-              val nextEdge = LoopEdge(b, nextNode)
-              var ih = nextNode
-
-              while (iloopHeaders.contains(ih)) {
-                ih = iloopHeaders(ih)
-                val thisLoop = loops(ih)
-                if (iloopHeaders.contains(from)) {
-                  if (iloopHeaders(from) != ih) {
-                    thisLoop.reducible = false
-                    thisLoop.reentries += edge
-                  }
-                } else {
-                  // `from` must be outside the loop
-                  thisLoop.reducible = false
-                  thisLoop.reentries += edge
-                }
+  private def traverse_loops_dfs(_istate: State, _b0: Block, _DFSPpos: Int): (State, Option[Block]) = {
```

Comment:
Nothing actually uses the second output parameter, what is it for & what does it mean?

Reply:
Do you mean the `Option[Block]`? I believe this is the first loop header following _b0, this is a holdover from the recursive style of the algorithm which traverses all the successors and tags block that is returned as a header. It is retained in the refactor so the psuedo-return  `retval` that is set in the `FinishProcessNode` state in the explicit call stack form matches the return value of the whole function.

Reply:
The only time traverse_loops_dfs gets called, it has the tuple access ._1 at the end which only returns the State. Do we need to have this second output parameter at all? If there is some reason we want to keep it, there should be a comment documenting what it means.

Reply:
I've just removed it as it isn't really necessary.